File: app/src/main/java/com/example/myapplication/core/catalog/AssetCatalogRepository.kt
package com.example.myapplication.core.catalog

import android.content.Context
import com.example.myapplication.core.model.ExerciseDefinition
import com.example.myapplication.core.model.ProgramTemplate

class AssetCatalogRepository(context: Context) {
    private val appContext = context.applicationContext

    val exercises: List<ExerciseDefinition> by lazy {
        CatalogParser.parseExercises(readAsset(EXERCISES_ASSET)).also { parsed ->
            val issues = CatalogValidator.validateExercises(parsed)
            check(issues.isEmpty()) {
                "Invalid bundled exercise catalog:\n${issues.joinToString(separator = "\n")}"
            }
        }
    }

    val programs: List<ProgramTemplate> by lazy {
        CatalogParser.parsePrograms(readAsset(PROGRAMS_ASSET))
    }

    private fun readAsset(path: String): String =
        appContext.assets.open(path).bufferedReader().use { it.readText() }

    private companion object {
        const val EXERCISES_ASSET = "catalog/exercises_vi.json"
        const val PROGRAMS_ASSET = "catalog/programs.json"
    }
}

File: flutter/lib/core/catalog/asset_catalog_repository.dart
import 'dart:async';
import 'catalog_parser.dart';
import 'catalog_validator.dart';
import '../model/catalog_models.dart';
import '../model/movement_block_models.dart';

class AssetCatalogRepository {
  final Future<String> Function(String) assetReader;

  AssetCatalogRepository({required this.assetReader});

  List<ExerciseDefinition>? _exercises;
  List<ProgramTemplate>? _programs;
  List<MovementBlock>? _movementBlocks;

  Future<void> init() async {
    if (_exercises != null && _programs != null && _movementBlocks != null) {
      return;
    }

    final exercisesRaw = await assetReader(_exercisesAsset);
    final parsedExercises = CatalogParser.parseExercises(exercisesRaw);
    final exerciseIssues = CatalogValidator.validateExercises(parsedExercises);
    if (exerciseIssues.isNotEmpty) {
      throw StateError(
          "Invalid bundled exercise catalog:\n${exerciseIssues.join('\n')}");
    }
    _exercises = parsedExercises;

    final programsRaw = await assetReader(_programsAsset);
    final parsedPrograms = CatalogParser.parsePrograms(programsRaw);
    final exercisesById = {for (var e in parsedExercises) e.id: e};
    final programIssues =
        CatalogValidator.validatePrograms(parsedPrograms, exercisesById);
    if (programIssues.isNotEmpty) {
      throw StateError(
          "Invalid bundled program catalog:\n${programIssues.join('\n')}");
    }
    _programs = parsedPrograms;

    final blocksRaw = await assetReader(_movementBlocksAsset);
    final parsedBlocks = CatalogParser.parseMovementBlocks(blocksRaw);
    final blockIssues = CatalogValidator.validateMovementBlocks(parsedBlocks);
    if (blockIssues.isNotEmpty) {
      throw StateError(
          "Invalid bundled movement blocks:\n${blockIssues.join('\n')}");
    }
    _movementBlocks = parsedBlocks;
  }

  List<ExerciseDefinition> get exercises =>
      _exercises ?? (throw StateError('Repository not initialized'));
  List<ProgramTemplate> get programs =>
      _programs ?? (throw StateError('Repository not initialized'));
  List<MovementBlock> get movementBlocks =>
      _movementBlocks ?? (throw StateError('Repository not initialized'));

  static const _exercisesAsset = 'assets/catalog/exercises_vi.json';
  static const _programsAsset = 'assets/catalog/programs.json';
  static const _movementBlocksAsset = 'assets/catalog/movement_blocks_vi.json';
}
